File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode f = head, s = head;
        while (true) {
            if (f == null || f.next == null) return null;
            f = f.next.next;
            s = s.next;
            if (f == s) break;
        }
        s = head;
        while (s != f) {
            f = f.next;
            s = s.next;
        }
        return s;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
class Solution {
public:
    ListNode* detectCycle(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return nullptr;
        }

        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* entry = head; 

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (fast == slow) {
                while(entry!=slow){
                    slow=slow->next;
                    entry=entry->next;
                }
                return entry;
            }
        }

        return nullptr;
    }
};

